CMakeLists.txt

```diff
@@ -1,281 +1,294 @@
 endif()
 
-if (NOT HAVE_SYS_QUEUE)
-    configure_file (
-        ${PROJECT_SOURCE_DIR}/compat/sys/queue.h.in
-        ${PROJECT_BINARY_DIR}/compat/sys/queue.h)
+if(NOT HAVE_SYS_QUEUE)
+    configure_file(
+        ${CMAKE_CURRENT_SOURCE_DIR}/compat/sys/queue.h.in
+        ${CMAKE_CURRENT_BINARY_DIR}/compat/sys/queue.h)
+    list(APPEND compat_headers "${CMAKE_CURRENT_BINARY_DIR}/compat/sys/queue.h")
+endif()
 
-    include_directories(${PROJECT_BINARY_DIR}/compat)
+if(NOT HAVE_SYS_TREE OR NOT HAVE_SYS_QUEUE)
+    target_include_directories(evhtp PUBLIC
+      $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/compat>)
 endif()
 
-if (NOT HAVE_STRNDUP)
-    add_definitions(-DNO_STRNDUP)
+if(NOT HAVE_STRNDUP)
+    target_compile_definitions(evhtp PUBLIC -DNO_STRNDUP)
 endif()
 
-if (NOT HAVE_STRNLEN)
-    add_definitions(-DNO_STRNLEN)
+if(NOT HAVE_STRNLEN)
+    target_compile_definitions(evhtp PUBLIC -DNO_STRNLEN)
 endif()
 
-if (NOT HAVE_SYS_UN)
-    add_definitions(-DNO_SYS_UN)
-endif(NOT HAVE_SYS_UN)
+if(NOT HAVE_SYS_UN)
+    target_compile_definitions(evhtp PUBLIC -DNO_SYS_UN)
+endif()
 
-if (HOST_BIG_ENDIAN)
-    add_definitions(-DHOST_BIG_ENDIAN)
+if(HOST_BIG_ENDIAN)
+    target_compile_definitions(evhtp PUBLIC -DHOST_BIG_ENDIAN)
 endif()
 
 # Test 32/64 bits
 if("${CMAKE_SIZEOF_VOID_P}" EQUAL "8")
-    add_definitions(-DEVHTP_SYS_ARCH=64)
+    target_compile_definitions(evhtp PUBLIC -DEVHTP_SYS_ARCH=64)
 elseif("${CMAKE_SIZEOF_VOID_P}" EQUAL "4")
-    add_definitions(-DEVHTP_SYS_ARCH=32)
+    target_compile_definitions(evhtp PUBLIC -DEVHTP_SYS_ARCH=32)
 else()
     message(ERROR "Unable to determine architecture")
 endif()
 
-if (EVHTP_USE_DEFER_ACCEPT)
-    add_definitions(-DUSE_DEFER_ACCEPT)
-endif(EVHTP_USE_DEFER_ACCEPT)
-
-set (CMAKE_INCLUDE_CURRENT_DIR ON)
-
-find_package(LibEvent REQUIRED)
-
-if (NOT EVHTP_DISABLE_REGEX)
-    find_package(Oniguruma)
-
-    if (NOT ONIGURUMA_FOUND)
-        message(STATUS "${BoldRed}${CMAKE_BUILD_TYPE}ONIGURUMA NOT FOUND, DISABLING REGEX SUPPORT${ColourReset}")
-        set(EVHTP_DISABLE_REGEX ON)
-    endif()
-endif()
-
-if (NOT EVHTP_DISABLE_SSL)
-    find_package(OpenSSL)
-endif()
-
-if (OPENSSL_FOUND)
-    if (APPLE)
-        # darwin based hosts have deprecated normal openssl functions, which is
-        # annoying to see warnings, for now, just ignore them.
-        add_definitions(-Wno-deprecated-declarations)
-    endif()
-endif()
-
-
-if (EVHTP_USE_JEMALLOC)
-        set(LIBEVHTP_EXTERNAL_LIBS ${LIBEVHTP_EXTERNAL_LIBS} jemalloc)
+if(EVHTP_USE_DEFER_ACCEPT)
+    target_compile_definitions(evhtp PUBLIC -DUSE_DEFER_ACCEPT)
 endif()
 
-if (EVHTP_USE_TCMALLOC)
-        set(LIBEVHTP_EXTERNAL_LIBS ${LIBEVHTP_EXTERNAL_LIBS} tcmalloc)
+if(APPLE)
+    # Darwin based hosts have deprecated normal openssl functions, which is
+    # annoying to see warnings, for now, just ignore them.
+    target_compile_options(evhtp PUBLIC -Wno-deprecated-declarations)
```

Comment:
I don't think such flags like warnings should be set to `PUBLIC`. I guess it's local project specific and hence should be set to `PRIVATE`.

Reply:
Good idea